### PR TITLE
test(logql): Add log-selection query support to the logqltest DSL

### PR DESCRIPTION
**Pull request by joe-elliott**
Adds support for LogQL selection queries to the logqltest DSL.

Basic smoke tests have also been added. We can expand on these further in future PRs.

### CODE REVIEW

File: pkg/logql/internal/logqltest/testdata/log_selection.logqltest [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+eval range from 0 to 30s step 10s {app="foo"}
+  {app="foo"} "1st line" @ 10s
+  {app="foo"} "2nd line" @ 20s
+
+# A regex selector across streams returns one result stream per match.
+eval range from 0 to 30s step 10s {app=~"foo|bar"}
+  {app="foo"} "1st line" @ 10s
+  {app="foo"} "2nd line" @ 20s
+  {app="bar"} "noise"    @ 15s
```
Comment (pracucci) [inline pkg/logql/internal/logqltest/testdata/log_selection.logqltest]: [nit] The idea of the noise, as documented in the AGENTS.md, is that it should be always filtered out by the selectors used in the queries. For this case, I think you could just move the noise line to another log stream that is not selected.

File: pkg/logql/internal/logqltest/runner.go [hunk 2]
```diff
@@ -399,6 +424,67 @@ func compareMatrix(name string, cmd evalCmd, exp expectations, m promql.Matrix,
 	return nil
 }
 
+// compareStreams checks a log-selection result against the expected streams. Streams are
+// matched as a set keyed by label string (like vector/matrix series); the log lines within a
+// matched stream are compared as an exact, ordered sequence, since a stream's line order is
```
Comment (pracucci) [inline pkg/logql/internal/logqltest/runner.go:434]: Shouldn't the expected output be in global chronological (or anti-chronological) order, instead of a per-stream assertion?

File: pkg/logql/internal/logqltest/README.md [hunk 3]
```diff
@@ -62,13 +62,20 @@ eval range from <t0> to <t1> step <step> <logql>
 - Times (`<time>`, `<t0>`, `<t1>`, `<step>`) are Go durations offset from the script epoch.
 - Expected results follow on indented lines. The block ends at a blank line, a dedented line,
   or EOF.
+- `<step>` is required by the grammar even for a log-selection query, which has no notion of a
```
Comment (pracucci) [inline pkg/logql/internal/logqltest/README.md]: We should fix it. We should make `<step>` not required. When not set we don't specify it in the request. Then, when the query response is received, we check if the step was set. If the response is a matrix and no step was set then we fail the test because for such query the `<step>` was required.

**Review by pracucci — DISMISSED** [review]
Nice work! Overall LGTM. My main comment is about `step <step>` that should not be specified for log queries (it's awkward having to specify it and also misleading). I left a suggestion on how to change it. I'm fine if you want to merge this PR as is, and then address it in another PR.

File: pkg/logql/internal/logqltest/README.md (hunk 3 above)
Comment (joe-elliott) [reply pkg/logql/internal/logqltest/README.md]: i tried a few approaches and liked this one the best. i opted for adding a new `select` keyword to cleanly disambiguate the three types of queries: `instant`, `range` and `select`. this will also make adding selection specific features like "direction" easier in the future. i can add this in a follow up PR.

https://github.com/grafana/loki/pull/24123/commits/6fbc730c19d30432d6464109d09dbcd382e13b21

wdyt?

File: pkg/logql/internal/logqltest/runner.go (hunk 2 above)
Comment (joe-elliott) [reply pkg/logql/internal/logqltest/runner.go:434]: the logql return value is grouped by stream and ordered within the stream:

https://github.com/grafana/loki/blob/main/pkg/push/types.go#L16-L28

so grouping the expected results by stream and comparing to the output of the engine seems preferred to flattening the actual output and comparing to the expected. if we sorted and flattened the output we'd be mutating the output which would weaken the value of the tests i think.

we could impose some requirements at parse time for readability? like the expected lines must be in overall time order regardless of the streams?

File: pkg/logql/internal/logqltest/testdata/log_selection.logqltest (hunk 1 above)
Comment (joe-elliott) [reply pkg/logql/internal/logqltest/testdata/log_selection.logqltest]: fixed in:

https://github.com/grafana/loki/pull/24123/changes/5354d7a3278d1f48e91faa6c1a8ffbd91d58dcb6

File: pkg/logql/internal/logqltest/runner.go (hunk 2 above)
Comment (pracucci) [reply pkg/logql/internal/logqltest/runner.go:434]: Thanks for the details. Let's keep it as is then.

File: pkg/logql/internal/logqltest/README.md (hunk 3 above)
Comment (pracucci) [reply pkg/logql/internal/logqltest/README.md]: I like `eval select`: it makes it explicit.

File: pkg/logql/internal/logqltest/testdata/AGENTS.md [hunk 4]
```diff
@@ -138,8 +138,19 @@ only what they can't show, and prefer a trailing `# …` on the relevant line ov
...
 
+## Log-selection scenarios
+
+A log-selection query (e.g. `{app="foo"} |= "bar"`) uses `eval select`, not `eval range`/`eval
+instant`, and returns streams, not series — see README.md "Streams" expected results. The
+instant/range cross-check in the checklist above doesn't apply the same way: `eval select` has no
+per-step matrix, just one window's worth of lines, so there's no "last step equals the instant" to
+size for. Instead cross-check by picking an `eval select` window and an `eval instant` time that
+should return the *same* lines (accounting for the default 30s look-back — see README.md
+"Log-selection window" for the exact boundary rule).
```
Comment (pracucci) [inline pkg/logql/internal/logqltest/testdata/AGENTS.md]: I think the `eval instant` mention below is wrong. That's for metrics, not log selection.

```suggestion
instant/range cross-check in the checklist above doesn't apply to log selection.
```

File: pkg/logql/internal/logqltest/exec.go [hunk 5]
```diff
@@ -44,7 +44,7 @@ type executionStack interface {
 
 // isQueryShardingSupported reports whether the shard mapper fans a query out into >= 2 shards.
 func isQueryShardingSupported(query string) bool {
-	expr, err := syntax.ParseSampleExpr(query)
+	expr, err := syntax.ParseExpr(query)
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/logql/internal/logqltest/exec.go:47]: I don't get the point. In practice, an instant query is always a metrics query. I think cursor is wrong here.

**Review by pracucci — DISMISSED** [review]
LGTM, modulo a comment on AGENTS.md (otherwise it may be deceived).

**Comment by joe-elliott** [comment]
> I don't get the point. In practice, an instant query is always a metrics query. I think cursor is wrong here.

Yes, Claude pointed this out to me as well and wrote some tests for it that I removed. There's an irrelevant edge case where an "instant logql selection" query is incorrectly marked as shardable by this code, but that kind of query doesn't exist. I have a note to dig a little deeper into it, but I don't think it's anything.

**Review by pracucci — APPROVED** [review]
